gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemIntegrationTest.java

```diff
@@ -1509,6 +1510,155 @@ public void testThreadTraceEnabledRename() throws Exception {
+
+      // delete /A/ non recursively
+      googleHadoopFileSystem.delete(new Path("gs://" + bucketName + "/A/"), false);
+
+      // create /C/ and rename it to /A/
+      googleHadoopFileSystem.mkdirs(new Path("C/"));
+
+      googleHadoopFileSystem.rename(
+          new Path("gs://" + bucketName + "/C/"), new Path("gs://" + bucketName + "/A/"));
+
+    } finally {
+      googleHadoopFileSystem.delete(new Path(initUri));
+    }
+  }
+
+  @Test
+  public void testHnBucketRecursiveDeleteOperationOnBucket() throws Exception {
+    String bucketName = this.gcsiHelper.getUniqueBucketName("hn");
+    GoogleHadoopFileSystem googleHadoopFileSystem = new GoogleHadoopFileSystem();
+    URI initUri = new URI("gs://" + bucketName);
+    Configuration config = loadConfig();
+    config.setBoolean("fs.gs.hierarchical.namespace.folders.enable", true);
+    googleHadoopFileSystem.initialize(initUri, config);
+
+    GoogleCloudStorage theGcs = googleHadoopFileSystem.getGcsFs().getGcs();
+    theGcs.createBucket(
+        bucketName, CreateBucketOptions.builder().setHierarchicalNamespaceEnabled(true).build());
+
+    assertThat(theGcs.isHnBucket(new Path(initUri.toString() + "/").toUri())).isTrue();
+
+    try {
+      googleHadoopFileSystem.mkdirs(new Path("A/"));
+      googleHadoopFileSystem.mkdirs(new Path("A/dir1/"));
+      googleHadoopFileSystem.mkdirs(new Path("A/dir2/"));
+      googleHadoopFileSystem.mkdirs(new Path("A/dir1/subdir1/"));
+      googleHadoopFileSystem.mkdirs(new Path("A/dir1/subdir2/"));
+      googleHadoopFileSystem.mkdirs(new Path("A/dir2/subdir3/"));
+      createFile(googleHadoopFileSystem, new Path("A/1"));
+      createFile(googleHadoopFileSystem, new Path("A/2"));
+
+      assertThrows(
+          "Cannot delete a non-empty directory",
+          java.nio.file.DirectoryNotEmptyException.class,
+          () -> googleHadoopFileSystem.delete(new Path("gs://" + bucketName + "/"), false));
+
+      // delete bucket
+      googleHadoopFileSystem.delete(new Path("gs://" + bucketName + "/"), true);
+
+      // create bucket with same name
+      theGcs.createBucket(
+          bucketName, CreateBucketOptions.builder().setHierarchicalNamespaceEnabled(true).build());
+      assertThat(theGcs.isHnBucket(new Path(initUri.toString() + "/").toUri())).isTrue();
+
+      // create the same folders
+      googleHadoopFileSystem.mkdirs(new Path("A/"));
+      googleHadoopFileSystem.mkdirs(new Path("A/dir1/"));
+      googleHadoopFileSystem.mkdirs(new Path("A/dir2/"));
+      googleHadoopFileSystem.mkdirs(new Path("A/dir1/subdir1/"));
+      googleHadoopFileSystem.mkdirs(new Path("A/dir1/subdir2/"));
+      googleHadoopFileSystem.mkdirs(new Path("A/dir2/subdir3/"));
+      createFile(googleHadoopFileSystem, new Path("A/1"));
+      createFile(googleHadoopFileSystem, new Path("A/2"));
+
+    } finally {
+      googleHadoopFileSystem.delete(new Path(initUri));
```

Comment:
where are we validating?

Reply:
This is deletion of entire bucket, we do not validate that. It is for cleaning purposes.

Reply:
We cannot validate that folder is deleted when the user commands bucket to be deleted. This test exists to make sure nothing breaks if user commands to delete a bucket. Test "testHnBucketRecursiveDeleteOperationOnDirectory" validates deletion of folder resource.

Reply:
Here we seem to be creating directories and files, but I am not seeing any validation that they are actually created.

Reply:
Added assertThat statements.